File: src/openharness/ui/mcp_notice.py
"""MCP 提示辅助函数，供 runtime、协议层和终端 UI 复用。"""

from __future__ import annotations

from collections.abc import Iterable

from openharness.mcp.types import McpConnectionStatus

_ERROR_TOKENS = (
    "fail",
    "error",
    "denied",
    "timeout",
    "timed out",
    "broken pipe",
    "disconnected",
    "not connected",
)
_SUCCESS_TOKENS = (
    "recover",
    "reconnected",
    "restored",
    "connected",
)


# 统一 MCP notice 的语义级别，避免多个 UI 模块各自猜测同一条状态文案。
_McpNoticeLevel = str


def infer_mcp_notice_level(state: str, detail: str, explicit_level: str | None = None) -> _McpNoticeLevel:
    """根据状态、详情和可选显式级别推断 MCP 提示语义。

    优先信任显式传入的级别；若没有，再结合运行时状态与 detail 文案兜底。
    """

    normalized_level = (explicit_level or "").strip().lower()
    if normalized_level in {"error", "failed", "failure"}:
        return "error"
    if normalized_level in {"success", "ok", "recovered"}:
        return "success"
    if normalized_level in {"warning", "warn", "notice", "info"}:
        return "warning"

    lowered = detail.lower()
    if state == "failed" or any(token in lowered for token in _ERROR_TOKENS):
        return "error"
    if state == "connected" or any(token in lowered for token in _SUCCESS_TOKENS):
        return "success"
    return "warning"


def summarize_mcp_notice_entry(statuses: Iterable[McpConnectionStatus]) -> tuple[str, str]:
    """提炼最近一次值得展示的 MCP 提示及其级别。"""

    status_list = list(statuses)
    for status in status_list:
        detail = (status.detail or "").strip()
        lowered = detail.lower()
        if detail and ("auto-reconnect" in lowered or "reconnect" in lowered or "recovered" in lowered):
            return detail, infer_mcp_notice_level(status.state, detail)
    for status in status_list:
        detail = (status.detail or "").strip()
        if status.state == "failed" and detail:
            return detail, "error"
    return "", ""


def format_mcp_notice_text(message: str, level: str | None = None) -> str:
    """把 MCP notice 格式化成纯文本标签，供 slash command 与摘要输出复用。"""

    normalized_level = infer_mcp_notice_level("pending", message, level)
    if normalized_level == "error":
        return f"mcp error: {message}"
    if normalized_level == "success":
        return f"mcp recovered: {message}"
    return f"mcp notice: {message}"


def format_mcp_notice_markup(message: str, level: str | None = None) -> str:
    """把 MCP notice 格式化成 Textual 可直接渲染的彩色 markup。"""

    normalized_level = infer_mcp_notice_level("pending", message, level)
    if normalized_level == "error":
        return f"[red]{format_mcp_notice_text(message, normalized_level)}[/red]"
    if normalized_level == "success":
        return f"[green]{format_mcp_notice_text(message, normalized_level)}[/green]"
    return f"[yellow]{format_mcp_notice_text(message, normalized_level)}[/yellow]"


def format_mcp_server_text_summary(statuses: Iterable[McpConnectionStatus]) -> str:
    """生成纯文本 MCP 摘要，显式标注每个 server 的详情语义。"""

    status_list = list(statuses)
    if not status_list:
        return "No MCP servers configured."

    lines = ["MCP servers:"]
    for status in status_list:
        lines.append(f"- {status.name} [{status.state}] {status.transport}")
        lines.append(
            f"  auth={bool(status.auth_configured)} tools={len(status.tools)} resources={len(status.resources)}"
        )
        if status.detail:
            lines.append(f"  {format_mcp_notice_text(status.detail, infer_mcp_notice_level(status.state, status.detail))}")
        if status.tools:
            lines.append(f"  tool_names: {', '.join(tool.name for tool in status.tools)}")
        if status.resources:
            lines.append(f"  resource_uris: {', '.join(resource.uri for resource in status.resources)}")
    return "\n".join(lines)


def format_mcp_server_panel_markup(statuses: Iterable[McpConnectionStatus]) -> str:
    """生成 Textual MCP 面板的逐 server markup 文本。

    每个 server 保留原始连接信息，并对 detail 行补充语义着色，方便快速定位
    哪个 server 刚恢复、哪个 server 仍然失败。
    """

    status_list = list(statuses)
    if not status_list:
        return "[b]MCP[/b]\n(none)"

    lines = ["[b]MCP[/b]"]
    for status in status_list[:5]:
        lines.append(f"{status.name} [{status.state}] {status.transport}")
        lines.append(
            f"[dim]auth={bool(status.auth_configured)} tools={len(status.tools)} resources={len(status.resources)}[/dim]"
        )
        if status.detail:
            lines.append(format_mcp_notice_markup(status.detail, infer_mcp_notice_level(status.state, status.detail)))
    return "\n".join(lines)


__all__ = [
    "format_mcp_notice_markup",
    "format_mcp_notice_text",
    "format_mcp_server_panel_markup",
    "format_mcp_server_text_summary",
    "infer_mcp_notice_level",
    "summarize_mcp_notice_entry",
]

File: frontend/terminal/src/components/mcpNotice.ts
/**
 * 表示 MCP 提示在终端中的语义级别，用于统一失败、恢复与普通提示的视觉反馈。
 */
export type McpNoticeTone = 'error' | 'success' | 'warning';

/**
 * 描述 MCP 提示在 React Terminal 中的展示结果，避免多个组件重复解析同一条文案。
 */
export type McpNoticePresentation = {
	message: string;
	tone: McpNoticeTone;
	label: string;
	color: 'red' | 'green' | 'yellow';
};

const ERROR_KEYWORDS = ['fail', 'error', 'denied', 'timeout', 'timed out', 'broken pipe', 'disconnected', 'not connected'];
const SUCCESS_KEYWORDS = ['recover', 'reconnected', 'connected', 'restored'];

/**
 * 解析 MCP 提示文案，给终端组件提供统一的语义标签与颜色。
 * 优先信任后端显式下发的 severity，缺失时再回退到关键字推断。
 */
export function describeMcpNotice(rawNotice: unknown, rawLevel?: unknown): McpNoticePresentation | null {
	const message = String(rawNotice ?? '').trim();
	if (!message) {
		return null;
	}

	const explicitTone = normalizeMcpNoticeTone(rawLevel);
	if (explicitTone) {
		return buildMcpNoticePresentation(message, explicitTone);
	}

	const normalized = message.toLowerCase();
	if (containsKeyword(normalized, ERROR_KEYWORDS)) {
		return buildMcpNoticePresentation(message, 'error');
	}

	if (containsKeyword(normalized, SUCCESS_KEYWORDS)) {
		return buildMcpNoticePresentation(message, 'success');
	}

	return buildMcpNoticePresentation(message, 'warning');
}

/**
 * 规范化后端传来的 severity 字段，兼容显式 level 与旧关键字别名。
 */
function normalizeMcpNoticeTone(rawLevel: unknown): McpNoticeTone | null {
	const normalized = String(rawLevel ?? '').trim().toLowerCase();
	if (!normalized) {
		return null;
	}
	if (['error', 'failed', 'failure'].includes(normalized)) {
		return 'error';
	}
	if (['success', 'ok', 'recovered'].includes(normalized)) {
		return 'success';
	}
	if (['warning', 'warn', 'notice', 'info'].includes(normalized)) {
		return 'warning';
	}
	return null;
}

/**
 * 根据统一 tone 生成终端展示对象，避免多个分支重复维护标签与颜色。
 */
function buildMcpNoticePresentation(message: string, tone: McpNoticeTone): McpNoticePresentation {
	if (tone === 'error') {
		return {
			message,
			tone,
			label: 'mcp error',
			color: 'red',
		};
	}
	if (tone === 'success') {
		return {
			message,
			tone,
			label: 'mcp recovered',
			color: 'green',
		};
	}
	return {
		message,
		tone,
		label: 'mcp notice',
		color: 'yellow',
	};
}

/**
 * 使用关键字匹配，避免在多个组件里重复维护同一套状态判断逻辑。
 */
function containsKeyword(message: string, keywords: string[]): boolean {
	return keywords.some((keyword) => message.includes(keyword));
}
